File: rust/skialin-core/src/paragraph.rs
use crate::{sys, Canvas};

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Affinity {
    Upstream,
    Downstream,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct GlyphPosition {
    pub position: i32,
    pub affinity: Affinity,
}

#[derive(Debug, Clone, PartialEq)]
pub struct GlyphInfo {
    pub bounds: crate::Rect,
    pub grapheme_cluster_range: std::ops::Range<usize>,
    pub direction: crate::TextDirection,
    pub is_ellipsis: bool,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct LineMetrics {
    pub start_index: usize,
    pub end_index: usize,
    pub end_excluding_whitespaces: usize,
    pub end_including_newline: usize,
    pub hard_break: bool,
    pub ascent: f64,
    pub descent: f64,
    pub unscaled_ascent: f64,
    pub height: f64,
    pub width: f64,
    pub left: f64,
    pub baseline: f64,
    pub line_number: usize,
}

pub struct Paragraph(pub(crate) *mut sys::skia::textlayout::Paragraph);

impl Paragraph {
    pub(crate) unsafe fn from_raw(ptr: *mut sys::skia::textlayout::Paragraph) -> Self {
        Paragraph(ptr)
    }

    pub fn layout(&mut self, width: f32) {
        unsafe { sys::skialin_bridge_Paragraph_layout(self.0, width) };
    }

    pub fn paint(&mut self, canvas: &mut Canvas, x: f32, y: f32) {
        unsafe { sys::skialin_bridge_Paragraph_paint(self.0, canvas.as_raw(), x, y) };
    }

    pub fn max_width(&self) -> f32 {
        unsafe { sys::skialin_bridge_Paragraph_getMaxWidth(self.0) }
    }

    pub fn height(&self) -> f32 {
        unsafe { sys::skialin_bridge_Paragraph_getHeight(self.0) }
    }

    pub fn min_intrinsic_width(&self) -> f32 {
        unsafe { sys::skialin_bridge_Paragraph_getMinIntrinsicWidth(self.0) }
    }

    pub fn max_intrinsic_width(&self) -> f32 {
        unsafe { sys::skialin_bridge_Paragraph_getMaxIntrinsicWidth(self.0) }
    }

    pub fn alphabetic_baseline(&self) -> f32 {
        unsafe { sys::skialin_bridge_Paragraph_getAlphabeticBaseline(self.0) }
    }

    pub fn ideographic_baseline(&self) -> f32 {
        unsafe { sys::skialin_bridge_Paragraph_getIdeographicBaseline(self.0) }
    }

    pub fn longest_line(&self) -> f32 {
        unsafe { sys::skialin_bridge_Paragraph_getLongestLine(self.0) }
    }

    pub fn did_exceed_max_lines(&self) -> bool {
        unsafe { sys::skialin_bridge_Paragraph_didExceedMaxLines(self.0) }
    }

    pub fn line_number(&mut self) -> usize {
        unsafe { sys::skialin_bridge_Paragraph_lineNumber(self.0) }
    }

    pub fn unresolved_glyphs(&mut self) -> Option<i32> {
        let count = unsafe { sys::skialin_bridge_Paragraph_unresolvedGlyphs(self.0) };
        (count >= 0).then_some(count)
    }

    /// The codepoints skparagraph could not resolve to a glyph during shaping - the actual
    /// characters behind [`Self::unresolved_glyphs`]'s count, for font-fallback-registry lookups.
    pub fn unresolved_codepoints(&mut self) -> Vec<i32> {
        let count = unsafe { sys::skialin_bridge_Paragraph_unresolvedCodepoints(self.0, std::ptr::null_mut(), 0) };
        if count <= 0 {
            return Vec::new();
        }
        let mut buf = vec![0i32; count as usize];
        unsafe { sys::skialin_bridge_Paragraph_unresolvedCodepoints(self.0, buf.as_mut_ptr(), count) };
        buf
    }

    /// Invalidates cached layout state so the next [`Self::layout`] call redoes
    /// shaping/positioning instead of being a no-op for an unchanged width.
    pub fn mark_dirty(&mut self) {
        unsafe { sys::skialin_bridge_Paragraph_markDirty(self.0) };
    }

    pub fn glyph_position_at_coordinate(&mut self, dx: f32, dy: f32) -> GlyphPosition {
        let mut affinity = 0i32;
        let position = unsafe { sys::skialin_bridge_Paragraph_getGlyphPositionAtCoordinate(self.0, dx, dy, &mut affinity) };
        GlyphPosition { position, affinity: if affinity == 0 { Affinity::Upstream } else { Affinity::Downstream } }
    }

    pub fn word_boundary(&mut self, offset: u32) -> std::ops::Range<usize> {
        let (mut start, mut end) = (0usize, 0usize);
        unsafe { sys::skialin_bridge_Paragraph_getWordBoundary(self.0, offset, &mut start, &mut end) };
        start..end
    }

    pub fn line_metrics_at(&self, line_number: i32) -> Option<LineMetrics> {
        let (mut start_index, mut end_index, mut end_excluding_whitespaces, mut end_including_newline) = (0usize, 0usize, 0usize, 0usize);
        let mut hard_break = 0i32;
        let (mut ascent, mut descent, mut unscaled_ascent, mut height, mut width, mut left, mut baseline) = (0f64, 0f64, 0f64, 0f64, 0f64, 0f64, 0f64);
        let found = unsafe {
            sys::skialin_bridge_Paragraph_getLineMetricsAt(
                self.0,
                line_number,
                &mut start_index,
                &mut end_index,
                &mut end_excluding_whitespaces,
                &mut end_including_newline,
                &mut hard_break,
                &mut ascent,
                &mut descent,
                &mut unscaled_ascent,
                &mut height,
                &mut width,
                &mut left,
                &mut baseline,
            )
        };
        found.then_some(LineMetrics {
            start_index,
            end_index,
            end_excluding_whitespaces,
            end_including_newline,
            hard_break: hard_break != 0,
            ascent,
            descent,
            unscaled_ascent,
            height,
            width,
            left,
            baseline,
            line_number: line_number as usize,
        })
    }

    pub fn line_metrics(&mut self) -> Vec<LineMetrics> {
        let count = self.line_number();
        (0..count as i32).filter_map(|i| self.line_metrics_at(i)).collect()
    }

    pub fn rects_for_range(&mut self, start: u32, end: u32, height_style: RectHeightStyle, width_style: RectWidthStyle) -> Vec<TextBox> {
        let count = unsafe { sys::skialin_bridge_Paragraph_getRectsForRange(self.0, start, end, height_style.into(), width_style.into(), std::ptr::null_mut(), 0) };
        if count <= 0 {
            return Vec::new();
        }
        let mut buf = vec![0f32; count as usize * 5];
        unsafe { sys::skialin_bridge_Paragraph_getRectsForRange(self.0, start, end, height_style.into(), width_style.into(), buf.as_mut_ptr(), count) };
        buf.chunks_exact(5)
            .map(|c| TextBox {
                rect: crate::Rect::new(c[0], c[1], c[2], c[3]),
                direction: if c[4] > 0.5 { crate::TextDirection::Ltr } else { crate::TextDirection::Rtl },
            })
            .collect()
    }

    pub fn rects_for_placeholders(&mut self) -> Vec<TextBox> {
        let count = unsafe { sys::skialin_bridge_Paragraph_getRectsForPlaceholders(self.0, std::ptr::null_mut(), 0) };
        if count <= 0 {
            return Vec::new();
        }
        let mut buf = vec![0f32; count as usize * 5];
        unsafe { sys::skialin_bridge_Paragraph_getRectsForPlaceholders(self.0, buf.as_mut_ptr(), count) };
        buf.chunks_exact(5)
            .map(|c| TextBox {
                rect: crate::Rect::new(c[0], c[1], c[2], c[3]),
                direction: if c[4] > 0.5 { crate::TextDirection::Ltr } else { crate::TextDirection::Rtl },
            })
            .collect()
    }

    pub fn glyph_info_at_utf16_offset(&mut self, code_unit_index: usize) -> Option<GlyphInfo> {
        let mut bounds = [0f32; 4];
        let (mut range_start, mut range_end) = (0usize, 0usize);
        let mut direction = 0i32;
        let mut is_ellipsis = false;
        let found = unsafe {
            sys::skialin_bridge_Paragraph_getGlyphInfoAtUTF16Offset(self.0, code_unit_index, bounds.as_mut_ptr(), &mut range_start, &mut range_end, &mut direction, &mut is_ellipsis)
        };
        found.then_some(GlyphInfo {
            bounds: crate::Rect::new(bounds[0], bounds[1], bounds[2], bounds[3]),
            grapheme_cluster_range: range_start..range_end,
            direction: if direction != 0 { crate::TextDirection::Ltr } else { crate::TextDirection::Rtl },
            is_ellipsis,
        })
    }

    pub fn closest_glyph_info_at(&mut self, dx: f32, dy: f32) -> Option<GlyphInfo> {
        let mut bounds = [0f32; 4];
        let (mut range_start, mut range_end) = (0usize, 0usize);
        let mut direction = 0i32;
        let mut is_ellipsis = false;
        let found = unsafe { sys::skialin_bridge_Paragraph_getClosestUTF16GlyphInfoAt(self.0, dx, dy, bounds.as_mut_ptr(), &mut range_start, &mut range_end, &mut direction, &mut is_ellipsis) };
        found.then_some(GlyphInfo {
            bounds: crate::Rect::new(bounds[0], bounds[1], bounds[2], bounds[3]),
            grapheme_cluster_range: range_start..range_end,
            direction: if direction != 0 { crate::TextDirection::Ltr } else { crate::TextDirection::Rtl },
            is_ellipsis,
        })
    }

    pub fn update_font_size(&mut self, from: usize, to: usize, font_size: f32) {
        unsafe { sys::skialin_bridge_Paragraph_updateFontSize(self.0, from, to, font_size) };
    }

    pub fn update_foreground_paint(&mut self, from: usize, to: usize, paint: &crate::Paint) {
        unsafe { sys::skialin_bridge_Paragraph_updateForegroundPaint(self.0, from, to, &*paint.0) };
    }

    pub fn update_background_paint(&mut self, from: usize, to: usize, paint: &crate::Paint) {
        unsafe { sys::skialin_bridge_Paragraph_updateBackgroundPaint(self.0, from, to, &*paint.0) };
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum RectHeightStyle {
    Tight,
    Max,
    IncludeLineSpacingMiddle,
    IncludeLineSpacingTop,
    IncludeLineSpacingBottom,
    Strut,
}

impl From<RectHeightStyle> for i32 {
    fn from(style: RectHeightStyle) -> Self {
        match style {
            RectHeightStyle::Tight => 0,
            RectHeightStyle::Max => 1,
            RectHeightStyle::IncludeLineSpacingMiddle => 2,
            RectHeightStyle::IncludeLineSpacingTop => 3,
            RectHeightStyle::IncludeLineSpacingBottom => 4,
            RectHeightStyle::Strut => 5,
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum RectWidthStyle {
    Tight,
    Max,
}

impl From<RectWidthStyle> for i32 {
    fn from(style: RectWidthStyle) -> Self {
        match style {
            RectWidthStyle::Tight => 0,
            RectWidthStyle::Max => 1,
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct TextBox {
    pub rect: crate::Rect,
    pub direction: crate::TextDirection,
}

impl Drop for Paragraph {
    fn drop(&mut self) {
        unsafe { sys::skialin_bridge_Paragraph_delete(self.0) };
    }
}

File: src/main/kotlin/org/skialin/Paragraph.kt
package org.skialin

import org.skialin.impl.Managed
import org.skialin.impl.NativeLoader

class Paragraph internal constructor(
    ptr: Long,
) : Managed(ptr, ParagraphNative::nRelease) {
    enum class Affinity { UPSTREAM, DOWNSTREAM }

    data class GlyphPosition(
        val position: Int,
        val affinity: Affinity,
    )

    data class GlyphInfo(
        val bounds: Rect,
        val graphemeClusterRange: LongRange,
        val direction: ParagraphStyle.TextDirection,
        val isEllipsis: Boolean,
    )

    data class LineMetrics(
        val startIndex: Long,
        val endIndex: Long,
        val endExcludingWhitespaces: Long,
        val endIncludingNewline: Long,
        val hardBreak: Boolean,
        val ascent: Double,
        val descent: Double,
        val unscaledAscent: Double,
        val height: Double,
        val width: Double,
        val left: Double,
        val baseline: Double,
    )

    fun layout(width: Float) = ParagraphNative.nLayout(nativePtr, width)

    fun paint(
        canvas: Canvas,
        x: Float,
        y: Float,
    ) = ParagraphNative.nPaint(nativePtr, canvas.ptr, x, y)

    val maxWidth: Float get() = ParagraphNative.nMaxWidth(nativePtr)
    val height: Float get() = ParagraphNative.nHeight(nativePtr)
    val minIntrinsicWidth: Float get() = ParagraphNative.nMinIntrinsicWidth(nativePtr)
    val maxIntrinsicWidth: Float get() = ParagraphNative.nMaxIntrinsicWidth(nativePtr)
    val alphabeticBaseline: Float get() = ParagraphNative.nAlphabeticBaseline(nativePtr)
    val ideographicBaseline: Float get() = ParagraphNative.nIdeographicBaseline(nativePtr)
    val longestLine: Float get() = ParagraphNative.nLongestLine(nativePtr)
    val didExceedMaxLines: Boolean get() = ParagraphNative.nDidExceedMaxLines(nativePtr)
    val lineNumber: Long get() = ParagraphNative.nLineNumber(nativePtr)

    val unresolvedGlyphs: Int?
        get() = ParagraphNative.nUnresolvedGlyphs(nativePtr).takeIf { it >= 0 }

    /** The codepoints skparagraph could not resolve to a glyph during shaping - the actual
     * characters behind [unresolvedGlyphs]'s count, for font-fallback-registry lookups. */
    val unresolvedCodepoints: IntArray
        get() = ParagraphNative.nUnresolvedCodepoints(nativePtr)

    /** Invalidates cached layout state so the next [layout] call redoes shaping/positioning
     * instead of being a no-op for an unchanged width. */
    fun markDirty() = ParagraphNative.nMarkDirty(nativePtr)

    fun glyphPositionAtCoordinate(
        dx: Float,
        dy: Float,
    ): GlyphPosition {
        val packed = ParagraphNative.nGlyphPositionAtCoordinate(nativePtr, dx, dy)
        return GlyphPosition((packed shr 32).toInt(), if ((packed and 1) == 0L) Affinity.UPSTREAM else Affinity.DOWNSTREAM)
    }

    fun wordBoundary(offset: Int): LongRange {
        val range = ParagraphNative.nWordBoundary(nativePtr, offset)
        return range[0] until range[1]
    }

    fun lineMetricsAt(lineNumber: Int): LineMetrics? {
        val out = DoubleArray(12)
        if (!ParagraphNative.nLineMetricsAt(nativePtr, lineNumber, out)) return null
        return LineMetrics(
            out[0].toLong(),
            out[1].toLong(),
            out[2].toLong(),
            out[3].toLong(),
            out[4] != 0.0,
            out[5],
            out[6],
            out[7],
            out[8],
            out[9],
            out[10],
            out[11],
        )
    }

    fun lineMetrics(): List<LineMetrics> = (0 until lineNumber.toInt()).mapNotNull { lineMetricsAt(it) }

    private fun flatToTextBoxes(flat: FloatArray): List<TextBox> =
        (0 until flat.size / 5).map { i ->
            val direction = if (flat[i * 5 + 4] > 0.5f) ParagraphStyle.TextDirection.LTR else ParagraphStyle.TextDirection.RTL
            TextBox(Rect(flat[i * 5], flat[i * 5 + 1], flat[i * 5 + 2], flat[i * 5 + 3]), direction)
        }

    fun getRectsForRange(
        start: Int,
        end: Int,
        heightStyle: RectHeightStyle = RectHeightStyle.TIGHT,
        widthStyle: RectWidthStyle = RectWidthStyle.TIGHT,
    ): List<TextBox> = flatToTextBoxes(ParagraphNative.nGetRectsForRange(nativePtr, start, end, heightStyle.ordinal, widthStyle.ordinal))

    fun getRectsForPlaceholders(): List<TextBox> = flatToTextBoxes(ParagraphNative.nGetRectsForPlaceholders(nativePtr))

    private fun readGlyphInfo(out: DoubleArray): GlyphInfo =
        GlyphInfo(
            Rect(out[0].toFloat(), out[1].toFloat(), out[2].toFloat(), out[3].toFloat()),
            out[4].toLong() until out[5].toLong(),
            if (out[6] > 0.5) ParagraphStyle.TextDirection.LTR else ParagraphStyle.TextDirection.RTL,
            out[7] > 0.5,
        )

    fun glyphInfoAt(codeUnitIndex: Long): GlyphInfo? {
        val out = DoubleArray(8)
        if (!ParagraphNative.nGlyphInfoAtUTF16Offset(nativePtr, codeUnitIndex, out)) return null
        return readGlyphInfo(out)
    }

    fun closestGlyphInfoAt(
        dx: Float,
        dy: Float,
    ): GlyphInfo? {
        val out = DoubleArray(8)
        if (!ParagraphNative.nClosestGlyphInfoAt(nativePtr, dx, dy, out)) return null
        return readGlyphInfo(out)
    }

    fun updateFontSize(
        from: Long,
        to: Long,
        fontSize: Float,
    ) = ParagraphNative.nUpdateFontSize(nativePtr, from, to, fontSize)

    fun updateForegroundPaint(
        from: Long,
        to: Long,
        paint: Paint,
    ) = ParagraphNative.nUpdateForegroundPaint(nativePtr, from, to, paint.nativePtr)

    fun updateBackgroundPaint(
        from: Long,
        to: Long,
        paint: Paint,
    ) = ParagraphNative.nUpdateBackgroundPaint(nativePtr, from, to, paint.nativePtr)
}

private object ParagraphNative {
    init {
        NativeLoader.ensureLoaded()
    }

    external fun nRelease(ptr: Long)

    external fun nLayout(
        ptr: Long,
        width: Float,
    )

    external fun nPaint(
        ptr: Long,
        canvasPtr: Long,
        x: Float,
        y: Float,
    )

    external fun nMaxWidth(ptr: Long): Float

    external fun nHeight(ptr: Long): Float

    external fun nMinIntrinsicWidth(ptr: Long): Float

    external fun nMaxIntrinsicWidth(ptr: Long): Float

    external fun nAlphabeticBaseline(ptr: Long): Float

    external fun nIdeographicBaseline(ptr: Long): Float

    external fun nLongestLine(ptr: Long): Float

    external fun nDidExceedMaxLines(ptr: Long): Boolean

    external fun nLineNumber(ptr: Long): Long

    external fun nUnresolvedGlyphs(ptr: Long): Int

    external fun nUnresolvedCodepoints(ptr: Long): IntArray

    external fun nMarkDirty(ptr: Long)

    external fun nGlyphPositionAtCoordinate(
        ptr: Long,
        dx: Float,
        dy: Float,
    ): Long

    external fun nWordBoundary(
        ptr: Long,
        offset: Int,
    ): LongArray

    external fun nLineMetricsAt(
        ptr: Long,
        lineNumber: Int,
        out: DoubleArray,
    ): Boolean

    external fun nGetRectsForRange(
        ptr: Long,
        start: Int,
        end: Int,
        heightStyle: Int,
        widthStyle: Int,
    ): FloatArray

    external fun nGetRectsForPlaceholders(ptr: Long): FloatArray

    external fun nGlyphInfoAtUTF16Offset(
        ptr: Long,
        codeUnitIndex: Long,
        out: DoubleArray,
    ): Boolean

    external fun nClosestGlyphInfoAt(
        ptr: Long,
        dx: Float,
        dy: Float,
        out: DoubleArray,
    ): Boolean

    external fun nUpdateFontSize(
        ptr: Long,
        from: Long,
        to: Long,
        fontSize: Float,
    )

    external fun nUpdateForegroundPaint(
        ptr: Long,
        from: Long,
        to: Long,
        paintPtr: Long,
    )

    external fun nUpdateBackgroundPaint(
        ptr: Long,
        from: Long,
        to: Long,
        paintPtr: Long,
    )
}
